zookeeper-server/src/test/java/org/apache/zookeeper/test/BackupManagerTest.java

```diff
@@ -0,0 +1,601 @@
+  }
+
+  private void setupZk() throws Exception {
+    LOG.info("Setting up directories");
+
+    dataDir = ClientBase.createTmpDir();
+    backupStatusDir = ClientBase.createTmpDir();
+    backupTmpDir = ClientBase.createTmpDir();
+    backupDir = ClientBase.createTmpDir();
+    backupStorage = new LocalBackupStorage(backupConfig);
+    ClientBase.setupTestEnv();
+
+    startZk();
+    snapLog = new FileTxnSnapLog(dataDir, dataDir);
+  }
+
+  private void startZk() throws Exception {
+    LOG.info("Starting Zk");
+
+    zks = new ZooKeeperServer(dataDir, dataDir, 3000);
+    SyncRequestProcessor.setSnapCount(100);
+    final int PORT = Integer.parseInt(HOSTPORT.split(":")[1]);
+    serverCnxnFactory = ServerCnxnFactory.createFactory(PORT, -1);
+    serverCnxnFactory.startup(zks);
+
+    LOG.info("Waiting for server startup");
+
+    Assert.assertTrue("waiting for server being up ",
+        ClientBase.waitForServerUp(HOSTPORT, CONNECTION_TIMEOUT));
+
+    backupManager = new BackupManager(dataDir, dataDir, backupStatusDir, backupTmpDir, 15,
+        backupStorage);
+    backupManager.initialize();
+    backupStatus = new BackupStatus(backupStatusDir);
+  }
+
+  private void closeZk() throws Exception {
+    LOG.info("Closing Zk");
+
+    if (serverCnxnFactory != null) {
+      serverCnxnFactory.closeAll(ServerCnxn.DisconnectReason.SERVER_SHUTDOWN);
+      serverCnxnFactory.shutdown();
+      serverCnxnFactory = null;
+    }
+
+    if (zks != null) {
+      zks.getZKDatabase().close();
+      zks.shutdown();
+      zks = null;
+    }
+
+    Assert.assertTrue("waiting for server to shutdown",
+        ClientBase.waitForServerDown(HOSTPORT, CONNECTION_TIMEOUT));
+  }
+
+  private void cleanupZk() throws Exception {
+    LOG.info("Closing and cleaning up Zk");
+
+    closeZk();
+    backupManager = null;
+    backupStatus = null;
+    serverCnxnFactory = null;
+    zks = null;
+    snapLog = null;
+  }
+
+  private ZooKeeper createConnection() throws IOException {
+    return new ZooKeeper(HOSTPORT, CONNECTION_TIMEOUT, this);
+  }
+
+  private void closeConnection(ZooKeeper zk) throws InterruptedException {
+    if (zk != null) {
+      zk.close();
+    }
+  }
+
+  private void createNodes(ZooKeeper zk, int count) throws Exception {
+    int last = createdNodeCount + count;
+    for (; createdNodeCount < last; createdNodeCount++) {
+      zk.create("/invalidsnap-" + createdNodeCount, new byte[0], Ids.OPEN_ACL_UNSAFE,
+          CreateMode.PERSISTENT);
+    }
+  }
+
+  private void createNode(ZooKeeper zk, String path) throws Exception {
+    zk.create(path, new byte[0], Ids.OPEN_ACL_UNSAFE, CreateMode.PERSISTENT);
+  }
+
+  private void deleteNode(ZooKeeper zk, String path) throws Exception {
+    zk.delete(path, -1);
+  }
+
+  private boolean exists(ZooKeeper zk, String path) throws Exception {
+    return zk.exists(path, false) != null;
+  }
+
+  private String[] getBackupFiles(File path) {
+    FilenameFilter filter = new FilenameFilter() {
```

Comment:
@mgao0 Upon giving this a second thought, do you think we should store them all in the same directory and not make the subfolders like /log or /snapshot? How hard do you think it would be to make this change?

This way we can keep this test structure (same as the Twitter design).

Unless you have a strong reason not to. Feel free to add your thoughts here.

Reply:
This is not hard to make the changes in storage end, I think a file name filter would work as if we have different subfolders for these two types of files. I personally think it will benefit us to store them into different subfolders though, it's more organized as in zookeeper server we also store log and snapshot in different subfolders, and also it's cleaner and more light weight than differentiating them using file name filter. How important it is for us to keep the test the same to the Twitter design? Is it ok that we divert from their implementation a little since this suits our use case more?

Reply:
Synced with @mgao0 offline.

We agreed to store both snapshot and Txlog backups in a flat directory instead of subfolders for the time being. This is to keep the testing logic even with the other open-source implementation. If we decide that this needs to change in the future, we will refactor the logic then.

The storage implementation will change, and the test will be left as is.